### PR TITLE
Add a --aot-lazy-assembly-load command line option to load assemblies…

### PR DESCRIPTION
**Pull request by vargaz**
… referenced by AOT images lazily.

Note that if this is enabled, and a referenced assembly is not found or its not
the exact version used to AOT the referencing assembly, the runtime will abort
because these kinds of errors can only be handled when loading the AOT image,
not later.

### CODE REVIEW

**Comment by vargaz** [comment]
https://github.com/dotnet/runtime/issues/66361

**Comment by vargaz** [comment]
This option can either be set on the command line, or from C code before
the runtime is initialized:
```
extern gboolean mono_opt_aot_lazy_assembly_load;

mono_opt_aot_lazy_assembly_load = TRUE;
```

**Comment by radical** [comment]
> This option can either be set on the command line, or from C code before the runtime is initialized:
> 
> ```
> extern gboolean mono_opt_aot_lazy_assembly_load;
> 
> mono_opt_aot_lazy_assembly_load = TRUE;
> ```

Does this need to be documented anywhere?

Will this be usable, and useful for wasm too? cc @lewing

**Review by lambdageek — DISMISSED** [review]

**Comment by lambdageek** [comment]
@vargaz these options are a little bit difficult for embedders to use - `mono_options_parse_options` isn't called from `mono_jit_parse_options()` or read from `MONO_ENV_OPTIONS`.

@steveisok is there some way to pass options to our Android template?  It would be nice if we could modify one of the functional tests to use this https://github.com/dotnet/runtime/tree/main/src/tests/FunctionalTests/Android/Device_Emulator

**Comment by steveisok** [comment]
> @steveisok is there some way to pass options to our Android template? It would be nice if we could modify one of the functional tests to use this https://github.com/dotnet/runtime/tree/main/src/tests/FunctionalTests/Android/Device_Emulator

We'd have to tweak a few things, but it shouldn't be too difficult.

**Comment by vargaz** [comment]
> @vargaz these options are a little bit difficult for embedders to use - `mono_options_parse_options` isn't called from `mono_jit_parse_options()` or read from `MONO_ENV_OPTIONS`.

Will fix that. I think it is parsed when parsing MONO_ENV_OPTIONS.

**Comment by vargaz** [comment]
Also, options.h could become part of the embedding api in the future if the api looks ok.

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/utils/options-def.h [hunk 1]
```diff
@@ -58,6 +58,7 @@ DEFINE_BOOL_READONLY(readonly_flag, "readonly-flag", FALSE, "Example")
 */
 
 DEFINE_BOOL(wasm_exceptions, "wasm-exceptions", FALSE, "Enable codegen for wasm exceptions")
+DEFINE_BOOL(aot_lazy_assembly_load, "aot-lazy-assembly-load", FALSE, "Load assemblies referenced by AOT images lazily")
```
Comment (lambdageek) [inline src/mono/mono/utils/options-def.h:61]: Should we make the default true for HOST_ANDROID, HOST_WASM, HOST_DARWIN && !HOST_OSX?

Comment (vargaz) [reply src/mono/mono/utils/options-def.h:61]: I think platforms should set it themselves.

Comment (vargaz) [reply src/mono/mono/utils/options-def.h:61]: Or maybe we can set it in the runtime startup code.

**Comment by eerhardt** [comment]
Have we tried using this on Maui Android and gotten any before/after startup perf numbers?

**Comment by lambdageek** [comment]
> Have we tried using this on Maui Android and gotten any before/after startup perf numbers?

@eerhardt we haven't tried

**Pull request merged**